mypy/checker.py

```diff
@@ -1486,7 +1491,12 @@ def check_multi_assignment_from_tuple(self, lvalues: List[Lvalue], rvalue: Expre
             if not undefined_rvalue:
                 # Infer rvalue again, now in the correct type context.
                 lvalue_type = self.lvalue_type_for_inference(lvalues, rvalue_type)
-                rvalue_type = cast(TupleType, self.expr_checker.accept(rvalue, lvalue_type))
+                reinferred_rvalue_type = self.expr_checker.accept(rvalue, lvalue_type)
+                if (isinstance(reinferred_rvalue_type, UnionType) and
```

Comment:
Can you add a comment here about what's going on? I think this may be the crucial place where Optional[X] becomes X?